src/computability/reduce.lean

```diff
@@ -0,0 +1,236 @@
+λ p q r ⟨f, c₁, i₁, h₁⟩ ⟨g, c₂, i₂, h₂⟩, 
+⟨g ∘ f, computable.comp c₂ c₁, injective_comp i₂ i₁, λ a,
+  ⟨λ h, by { rwa [←h₂, ←h₁] }, 
+   λ h, by { rwa [h₁, h₂] }⟩⟩
+
+namespace computable_pred
+variables {α : Type*} {σ : Type*}
+variables [primcodable α] [primcodable σ]
+open computable
+
+theorem computable_of_many_one_reducible
+  {p q : α → Prop} 
+  (h₁ : p ≤₀ q) (h₂ : computable_pred q) : computable_pred p := 
+  begin
+    rcases h₁ with ⟨f, c, hf⟩,
+    rcases computable_iff.1 h₂ with ⟨g, hg, rfl⟩, split, 
+    { apply computable.of_eq (cond (comp hg c) (const tt) (const ff)),
+      intro, cases heq : g (f n), repeat { simp [hf, heq] },
+      { intro, cases heq : g (f a), 
+        { left, simp [hf, heq] },
+        { right, simp [hf, heq] } } }
+  end
+
+theorem computable_of_one_one_reducible
+  {p q : α → Prop} 
+  (h₁ : p ≤₁ q) (h₂ : computable_pred q) : computable_pred p := 
+  by { rcases h₁ with ⟨f, c, hinj, hf⟩, apply computable_of_many_one_reducible ⟨f, c, hf⟩ h₂ }
+
+end computable_pred
+
+def many_one_equiv {α} [primcodable α] (p q : α → Prop) := p ≤₀ q ∧ q ≤₀ p
+
+def one_one_equiv {α} [primcodable α] (p q : α → Prop) := p ≤₁ q ∧ q ≤₁ p
+
+theorem equivalence_of_many_one_equiv {α} [primcodable α] : equivalence (@many_one_equiv α _) :=
+⟨λ x,⟨reflexive_many_one_reducible _, reflexive_many_one_reducible _⟩,  
+ λ x y h, ⟨h.2, h.1⟩, 
+ λ x y z h₁ h₂, ⟨transitive_many_one_reducible h₁.1 h₂.1, transitive_many_one_reducible h₂.2 h₁.2⟩⟩
+
+theorem equivalence_of_one_one_equiv {α} [primcodable α] : equivalence (@one_one_equiv α _) :=
+⟨λ x,⟨reflexive_one_one_reducible _, reflexive_one_one_reducible _⟩,  
+ λ x y h, ⟨h.2, h.1⟩, 
+ λ x y z h₁ h₂, ⟨transitive_one_one_reducible h₁.1 h₂.1, transitive_one_one_reducible h₂.2 h₁.2⟩⟩
+
+instance many_one_equiv_setoid {α} [primcodable α] : setoid (set α) := 
+⟨many_one_equiv, @equivalence_of_many_one_equiv α _⟩
+
+def many_one_degree {α} [primcodable α] := quotient (@many_one_equiv_setoid α _)
+
+def disjoin (p q : set ℕ) : set ℕ  := (λ x, 2 * x) '' p ∪ (λ x, 2 * x + 1) '' q
+
+local infix ` ⊕ `:1001 := disjoin
+
+theorem mem_disjoin_left {p q : set ℕ} {n : ℕ} (h₁ : ¬nat.bodd n) (h₂ : (p ⊕ q) n) : p (nat.div2 n) :=
+begin
+  cases h₂,
+  { rcases h₂ with ⟨w, hl, hr⟩, rw ←hr, dsimp, rw ←nat.bit0_val, simpa [nat.div2_bit0] using hl },
+  { rcases h₂ with ⟨w, hl, hr⟩, rw ←hr at h₁, simp at h₁, contradiction }
+end
+
+theorem mem_disjoin_right {p q : set ℕ} {n : ℕ} (h₁ : nat.bodd n) (h₂ : (p ⊕ q) n) : q (nat.div2 n) :=
+begin
+  cases h₂,
+  { rcases h₂ with ⟨w, hl, hr⟩, rw ←hr at h₁, simp at h₁, contradiction },
+  { rcases h₂ with ⟨w, hl, hr⟩, rw ←hr, dsimp, rw ←nat.bit1_val, simpa [nat.div2_bit1] using hl }
+end
+
+namespace nat.primrec
+
+lemma add1 : primrec (nat.unpaired (λ x y, nat.succ y)) :=
+by simp [primrec₂.unpaired, primrec.comp₂ primrec.succ primrec₂.right]
+
+lemma double : primrec (λ n, 2 * n) :=
```

Comment:
I am concerned that this theorem is not a trivial consequence of the primrec library. It should be the composition of primrec_mul and primrec_const

Reply:
This one drives me crazy. I'm still not sure how to prove it elegantly. Apparently I haven't fully understood the primrec library :(

Reply:
you should not be using `nat.primrec`, that is only for bootstrapping. `primrec` also works on functions from nat to nat and has a lot more theorems